### PR TITLE
Reduce projections of type aliases with class type prefixes

### LINKED ISSUE #19892: Project compiles in 18 seconds on scala 2, compilation never ends on scala 3
## Compiler version

3.3.3 and 3.4.0

## Minimized code

The project I'm attempting to compile is [here](https://gitlab.com/joliciel/test-server).

```shell
git clone https://gitlab.com/joliciel/test-server.git
cd test-server
sbt
```
I removed the business modules (which compile quickly) and kept only the api module (whose compilation never ends).
I haven't minimized the code further as I could find no way of determining which parts were freezing the compiler.

## Output

```shell
 % sbt
[info] welcome to sbt 1.9.9 (Eclipse Adoptium Java 17.0.2)
[info] loading global plugins from /home/assaf/.sbt/1.0/plugins
[info] loading settings for project test-server-build from plugins.sbt ...
[info] loading project definition from /home/assaf/javawork/jochre3/test-server/project
[info] loading settings for project jochre3-server from build.sbt ...
[warn] JOCHRE3_SERVER_VERSION env var not found
[info] set current project to jochre3-server (in build file:/home/assaf/javawork/jochre3/test-server/)
[info] sbt server started at local:///home/assaf/.sbt/1.0/server/92affff4cbf1ff9d4214/sock
[info] started sbt server
sbt:jochre3-server> clean
[success] Total time: 0 s, completed Mar 6, 2024, 4:02:19 PM
sbt:jochre3-server> compile
[info] compiling 22 Scala sources to /home/assaf/javawork/jochre3/test-server/modules/api/target/scala-2.13/classes ...
[warn] 1 deprecation
[warn] 1 deprecation (since 3.7.0)
[warn] 2 deprecations in total; re-run with -deprecation for details
[warn] three warnings found
[success] Total time: 18 s, completed Mar 6, 2024, 4:02:40 PM
sbt:jochre3-server> ++ 3.3.3
[info] Setting Scala version to 3.3.3 on 1 projects.
[info] Excluded 1 projects, run ++ 3.3.3 -v for more details.
[info] Reapplying settings...
[info] set current project to jochre3-server (in build file:/home/assaf/javawork/jochre3/test-server/)
sbt:jochre3-server> compile
[info] compiling 22 Scala sources to /home/assaf/javawork/jochre3/test-server/modules/api/target/scala-3.3.3/classes ...

  | => api / Compile / compileIncremental 3292s
```

## Expectation

I would expect scala 3 to compile the code, which scala 2.13.13 compiles in 18 seconds, in a reasonable amount of time.

### PR DESCRIPTION
**Pull request by odersky**
Projections P # X are types that we would like to avoid. If X is a class type, there's nothing we can do. If X is an abstract type, we use skolemization and rewrite to (x?: P).X. If X is an alias type we should simply dealias but this was not done before.

This caused an exponential blowup in #19892, where we constructed types of the form

   ZPartialServerEndpoint[R, A, B, I, E, O, -C] # EndpointType[A, I, E, T, R] ... # EndpointType[A, I, E, T, R]

When the were 5 or more such selections, compile times blew up (33s for 5, timeout after 5 minutes for 6). I am still not quite sure where the blowup happened. Looking at stacktraces of random interrupts it seemed to be in a deep recursion of memberDenot and asSeenFrom calls.I believe it would still be interesting to find out more about this, in case there are other similar situations where combinations of deep projections with wide applications cannot be avoided.

But for this precise problem, eagerly dealiasing fixes it.

Fixes #19892

### CODE REVIEW

**Comment by odersky** [comment]
@rochala Can you help me fix the 2 presentation compiler test failures? I am not sure what to do here.

**Comment by rochala** [comment]
I'll check it the first thing in the morning

**Comment by rochala** [comment]
Dealiasing affects the typed trees which we are using for various features.
The tests that failed can be fixed by migrating to untyped trees, but there are other features such as hovers / go to type definition which can be more problematic, as the tree contains already dealiased type. 

I suppose it could be done if we analyse each select whether it is actually a type projection, and try recreating the original type by hand.

I will try to implement those workarounds unless you have a better idea.

**Comment by odersky** [comment]
Does it just apply to # projections `A # X`? In that case I think we can also simply drop the behavior and the tests. Such projections are quite rare. For instance the dotty code base does not have a single occurrence of #. So I personally would not care at all what kind of hovers I get. And for others it would be a good indicator to change the design, since IME relying on # usually does not end well. 

I believe the _only_ reason to use A # B is if both `A` and `B` are classes and we want to model references to inner classes in the Java style. Otherwise, if you have to resort to it, it usually means your design is wrong. Maybe we should do a linter rule for that.

**Comment by rochala** [comment]
> Does it just apply to # projections A # X?
It seems like this is the case.

This functionality was only tested in one suite, the one that failed, so I can say it wasn't truly supported by presentation compiler. This also impacts generated semanticdb output.

Feel free to add `@Ignore` for those tests. I'll create the ticket for this after this PR is merged so we don't lose a track of this. 
Other than that, I think the review of the solution should be done by someone more experienced.

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
We should get benchmarks back before we can merge this. I want to rule out slowdowns in some scenarios.

**Comment by odersky** [comment]
So performance seems to not worse and maybe a bit better. @smarter do you want to give this q quick review?

**Review by smarter — CHANGES_REQUESTED** [review]
Otherwise LGTM.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -2708,13 +2708,21 @@ object Types extends TypeUtils {
...
-    /** Reduce a type-ref `T { X = U; ... } # X`  to   `U`
-     *  provided `U` does not refer with a RecThis to the
-     *  refinement type `T { X = U; ... }`
+    /** Reduce a type ref P # X, where X is a type alias and P is a refined type or
+     *  a class type. If P is a refined type `T { X = U; ... }`, reduce P to U,
+     *  provided U does not refer with a RecThis to the same refined type. If P is a
+     *  class type, reduce it to the dealiasd version of P # X. This means that at typer
+     *  we create projections only for inner classes with class prefixes, since projections
+     *  of P # X where X is an abstract type are handled by skolemization. At later phases
+     *  these projections might arise, though.
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Types.scala:2717]: Yes, but I am not sure we can dealias without breaking stuff here. It is a super low-level code.

**Pull request merged**